### PR TITLE
set_artifact action plugin with tests

### PR DESCRIPTION
**Pull request by chrismeyersfsu**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

`set_artifact`

ansible/ansible-modules-core#5776

### CODE REVIEW

**Comment by abadger** [comment]
The new tests are failing.
needs_revision

File: test/integration/targets/set_artifact/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  set_artifact:
+    data:
+      b: 8
+  register: result_merge
+
+- name: Ensure set_artifacts does not set facts on hosts other than the first
+  assert:
+    that:
+      - artifact_data is undefined
+  when: "inventory_hostname != ansible_play_hosts[0]"
```
Comment (bcoca) [inline test/integration/targets/set_artifact/tasks/main.yml:49]: This test is irrelevant as the play only runs against 'testhost'

File: test/integration/targets/set_artifact/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+- name: Verify first-host artifact_data fact set
+  assert:
+    that:
+      - artifact_data is defined
+      - artifact_data.a == 7
+      - artifact_data.b == 8
+      - artifact_data|length == 2
+  run_once: true
```
Comment (bcoca) [inline test/integration/targets/set_artifact/tasks/main.yml:58]: run_once does not matter ... as play only uses single host

**Comment by amenonsen** [comment]
An interesting-looking new module PR… with *not even one line* of explanation about what it does?

**Comment by mattclay** [comment]
For tests that have special requirements, such as needing custom inventory, use a script based test instead of a role based test.

Take a look at the [unicode test](https://github.com/ansible/ansible/blob/devel/test/integration/targets/unicode/runme.sh) for an example.

**Comment by chrismeyersfsu** [comment]
@amenonsen since this is a action plugin there is a corresponding ansible module that contains the docs, see https://github.com/ansible/ansible-modules-core/pull/5776

@mattclay It's been a while since I've delved into the Ansible test structure. Previously, adding the role invocation to `non_destructive.yml` and forming your tests around the respective `inventory` was the way to go. Is `non_destructive.yml` not a thing anymore? Thanks for the `unicode test` reference.

**Comment by bcoca** [comment]
@chrismeyersfsu it is but non_destructive does not test 'multi host issues', which your tests seem to require.

**Comment by chrismeyersfsu** [comment]
yep, your right. I'll use the custom inventory methodology.

**Comment by bcoca** [comment]
-1, this seems like an alternate implementation of a fact/non_fact cache with aggregation of dicts, which can also be achieved via template/vars_fiiles|include_vars and combine filter.

**Comment by amenonsen** [comment]
As @bcoca says, I've found this functionality useful enough to implement it using include_vars (though I used copy with ``content: {{ variables|to_json }}`` instead of template; minor difference). So I like the idea, but I don't much like ``set_artifact``. It just feels a bit arbitrary and confusing. I know what an artifact is, but I don't know why I would be setting one, and the module documentation doesn't really help me. The term brings to mind build artifacts (e.g., ``*.pyc`` files) rather than anything else.

I wouldn't mind a ``set_variable`` that made it convenient to persistently set variables in the local inventory (and maybe even the remote fact cache), so maybe it's a matter of repackaging this functionality in a more general way.

**Pull request closed without merging**